router: retry on connect failure

fl0Lec opened the pull request:
Fix #32464


Commit Message: retry on connect failure
Additional Description: see bellow
Risk Level: High
Testing: unit test and build locally
Docs Changes: No
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

# More context
## *Description*:
When retrying on connection failure all requests having a body bigger than the connection buffer limit will not be retried.

Give the configuration bellow all requests bigger than 32768 bytes will not be retried. If not overridden per route it will come from the listener config `per_connection_buffer_limit_bytes` and default to [1MiB](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto)

The following happens:
1. the router receive the body and proxy it to the upstream requests
2. the buffer limit get reached at the router level
3. the router give up on retry but keep on forwarding to the upstream requests
4. the buffer of the upstream connection will get full and notify to stop readying on the downstream connection.

If the upstream requests fail the requests will not be retried.

## *Proposed solution*:
The router can a callback for when the upstream connection get a connection via the `addUpstreamCallbacks`.

If the router buffer get full before any upstream connection got establish keep on buffering above the limit and call the `onDecoderFilterAboveWriteBufferHighWatermark()` callback to ensure we stop on reading from the downstream. This allows to keep on retrying will keeping the buffer at a value close to the desired limit.

Once it get a new connection enable drain the buffer which enable read from downstream and drop retry, shadow, and buffering.

For all other retry the behaviour is unchanged

## *Example*:
config:
```yaml
static_resources:
  listeners:
  - name: test_connect_failure    
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 8100
    per_connection_buffer_limit_bytes: 32768
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: AUTO
          stat_prefix: web1
          route_config:
            virtual_hosts:
            - name: backend
              domains:
              - "*"
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: local_service_fail
                  timeout: 0s
                  retry_policy:
                    retry_on: "connect-failure"
                    num_retries: 3
                    retry_back_off:
                      base_interval: 0.050s
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
  clusters:
  - name: local_service_fail
    connect_timeout: 1s
    type: STRICT_DNS
    lb_policy: ROUND_ROBIN
    upstream_connection_options:
      tcp_keepalive:
        keepalive_time: 300
    common_lb_config:
      healthy_panic_threshold:
        value: 40
    load_assignment:
      cluster_name: local_service
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: 127.0.0.1
                # port doesn't match anything
                port_value: 8085
```

with Envoy docker image: "envoyproxy/envoy:v1.29.0":

only relevant logs:

```
[2024-02-13 09:25:46.619][21][debug][connection] [./source/common/network/connection_impl.h:98] [Tags: "ConnectionId":"0"] current connecting state: false
[2024-02-13 09:25:46.619][21][debug][router] [source/common/router/router.cc:514] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] cluster 'local_service_fail' match for URL '/'
[2024-02-13 09:25:46.619][21][debug][router] [source/common/router/router.cc:731] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] router decoding headers:
':authority', 'localhost:8100'
':path', '/error'
':method', 'POST'
':scheme', 'http'
'user-agent', 'Go-http-client/1.1'
'content-length', '92768'
'accept-encoding', 'gzip'
'x-forwarded-proto', 'http'
'x-request-id', '00db4f6a-b6cb-4fe3-b4d4-34bad1e002f2'

[2024-02-13 09:25:46.619][21][debug][connection] [./source/common/network/connection_impl.h:98] [Tags: "ConnectionId":"1"] current connecting state: true
[2024-02-13 09:25:46.619][21][debug][connection] [source/common/network/connection_impl.cc:1009] [Tags: "ConnectionId":"1"] connecting to 127.0.0.1:8085
[2024-02-13 09:25:46.619][21][debug][connection] [source/common/network/connection_impl.cc:1028] [Tags: "ConnectionId":"1"] connection in progress
[2024-02-13 09:25:46.619][21][trace][connection] [source/common/network/connection_impl.cc:654] [Tags: "ConnectionId":"0"] read ready. dispatch_buffered_data=0
[2024-02-13 09:25:46.619][21][trace][connection] [source/common/network/raw_buffer_socket.cc:25] [Tags: "ConnectionId":"0"] read returns: 32768
[2024-02-13 09:25:46.619][21][debug][router] [source/common/router/router.cc:856] The request payload has at least 65412 bytes data which exceeds buffer limit 32768. Give up on the retry/shadow.
[2024-02-13 09:25:46.619][21][debug][http] [source/common/http/filter_manager.cc:541] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] request data too large watermark exceeded
[2024-02-13 09:25:46.619][21][debug][http] [source/common/http/conn_manager_impl.cc:1922] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] Read-disabling downstream stream due to filter callbacks.
[2024-02-13 09:25:46.619][21][trace][connection] [source/common/network/connection_impl.cc:401] [Tags: "ConnectionId":"0"] readDisable: disable=true disable_count=0 state=0 buffer_length=0
[2024-02-13 09:25:46.619][21][debug][connection] [source/common/network/connection_impl.cc:757] [Tags: "ConnectionId":"1"] delayed connect error: 111
[2024-02-13 09:25:46.619][21][debug][connection] [source/common/network/connection_impl.cc:278] [Tags: "ConnectionId":"1"] closing socket: 0
[2024-02-13 09:25:46.619][21][trace][connection] [source/common/network/connection_impl.cc:469] [Tags: "ConnectionId":"1"] raising connection event 0
[2024-02-13 09:25:46.619][21][debug][router] [source/common/router/router.cc:1331] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] upstream reset: reset reason: remote connection failure, transport failure reason: delayed connect error: 111
[2024-02-13 09:25:46.619][21][debug][http] [source/common/http/conn_manager_impl.cc:1912] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] Read-enabling downstream stream due to filter callbacks.
[2024-02-13 09:25:46.619][21][trace][connection] [source/common/network/connection_impl.cc:401] [Tags: "ConnectionId":"0"] readDisable: disable=false disable_count=1 state=0 buffer_length=0
[2024-02-13 09:25:46.619][21][debug][http] [source/common/http/filter_manager.cc:1065] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] Sending local reply with details upstream_reset_before_response_started{remote_connection_failure,delayed_connect_error:_111}
[2024-02-13 09:25:46.619][21][debug][http] [source/common/http/conn_manager_impl.cc:1869] [Tags: "ConnectionId":"0","StreamId":"742379432996652670"] encoding headers via codec (end_stream=false):
':status', '503'
'content-length', '152'
'content-type', 'text/plain'
'date', 'Tue, 13 Feb 2024 09:25:46 GMT'
'server', 'envoy'
'connection', 'close'
```

# Test
- unit test
- integration test
- tested in one AZ of a staging environment, saw the UF response flags going away
![UF response flags (2)](https://github.com/envoyproxy/envoy/assets/104759954/383610e9-1384-459b-9a39-69802ea91ba2)

fl0Lec commented:
/retest

fl0Lec commented:
also opened an issue: https://github.com/envoyproxy/envoy/issues/32464

yanavlasov commented:
@fl0Lec thank you for the change. Can you describe the problem that you are trying to solve, please? I.e. is there an open issue you are fixing?

I think retries on TCP connection failures should already working. Are they not?

@alyssawilk for more comments.

/wait-any

alyssawilk commented:
+1 - I think it would help if you could write an integration test which simulates these conditions, fails without your change and succeeds with your change.  Mind giving that a shot?
/wait

fl0Lec commented:
Make sense, I forgot to add the link to the issue I created: https://github.com/envoyproxy/envoy/issues/32464
This add more context on the problem that I am trying to solve here.

I will look into adding an integration test

alyssawilk commented:
Ah, with the attached issue that makes much more sense, thanks so much.  So to summarize if we'd be willing to buffer X bytes in an upstream send buffer before applying flow control we should be willing to buffer X bytes (larger than the default of Y) in the retry buffer when we don't haven a upstream connection?   As long as we cancel the retry once we connect so we don't buffer the extra bytes for a long duration (which it looks like you're doing) I think that's sensible.   Definitely lets get e2e tests, ideally a bit more commenting, and runtime guard as well as it's a pretty high risk change.

fl0Lec commented:
I added per your suggestion @alyssawilk:
- runtime guard (defaulting to false as it is a pretty risky change)
- renamed some variable for clarity
- 2 integration tests:
	- 1 with the new behaviour (retry even when buffer limit exceeded)
	- 1 with the old behaviour (return 5xx)

I also update the PR description to give more context.
This PR has also currently been tested in part of our staging environment. I will like a graph in the PR comment soon

fl0Lec commented:
/retest

alyssawilk commented:
adding in @KBaichoo for first pass

KBaichoo commented on source/common/router/router.cc:
```diff
@@ -848,28 +882,42 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
+      buffer_limit_exceeded_no_connection_ = true;
+      // ensure we stop on reading, the buffer limit might be greater than
+      // retry_shadow_buffer_limit_ due to an override per route of the retry_shadow_buffer limit
+      callbacks_->onDecoderFilterAboveWriteBufferHighWatermark();
```
This does not necessarily hard stop data from entering the proxy, but it might stop this buffer from receiving more data. What is the limit on how much we're willing to buffer?

fl0Lec replied:
That make perfect sense, I hadn't fully understand Alyssa comment.
 
I update the code to do what you suggest. 
Setting a hard limit of two times the upstream connection buffer limit. 
In you example, we can not set the limit to Y, the idea would be to reach the limit on the upstream buffer in order to:
- pause the read on the downstream connection.
- not block data anywhere in the proxy as it might have been the case before.

Does that make sense to you ?
two times the upstream buffer limit can probably be discussed or coming from a runtime value with an appropriate default

KBaichoo commented on source/common/router/router.cc:
```diff
@@ -837,6 +842,35 @@ void Filter::sendNoHealthyUpstreamResponse() {
+  if (buffer_limit_exceeded_no_connection_) {
+    ENVOY_LOG(debug, "Giving up on retry and shadow and enabling read on downstream connection");
+    // disabling retry and shadow: the requests is to big
+    giveUpRetryAndShadow(true);
```
IIUC, are we just deferring giving up retry and shadow until either (1) out of retry attempts or (2) we have a connection established?

Prior we'd cancel retries if the shadow limit was reached immediately and send a local reply

KBaichoo commented on test/integration/integration_test.cc:
```diff
@@ -2409,6 +2409,103 @@ TEST_P(IntegrationTest, RandomPreconnect) {
+  codec_client_->sendData(encoder, 128, false);
+
+  // waiting to have at least one connection failure
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_connect_fail", 0);
```
shouldn't this wait for this to be 1?

fl0Lec replied:
I guess this could be a no-op in some case but it guarantee us that we got at least one connection failure before creating a healthy upstream. Which is what we want to test here.
IMO it does not make the test more flaky as we allows up to 5 retries, quite the opposite as I explained above

I ran the test around 10 times on my local machine without any issue

KBaichoo replied:
it guarantee us that we got at least one connection failure before creating a healthy upstream
> but the counter tests for >= 0, which by default it starts as 0?

fl0Lec replied:
Right make perfect sense, I thought it was strictly greater for some reason

KBaichoo reviewed (commented):
Thank you for working on this

/wait

KBaichoo reviewed (commented):
Thank you

/wait

KBaichoo commented on source/common/router/router.cc:
```diff
@@ -1976,6 +2011,9 @@ void Filter::doRetry(bool can_send_early_data, bool can_use_http3, TimeoutRetry
   UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && wait_for_connect_) {
+    upstream_request->addUpstreamCallbacks(*this);
```
do we still need to add to the upstream callbacks now?

fl0Lec replied:
Yes, otherwise how can we know that we got a connection with the upstream.
cf linge 862 above `onUpstreamConnectionEstablished`: https://github.com/DataDog/envoy/blob/c0e00f3a9f0d21b76a6d14c98efbc65341ed9790/source/common/router/router.cc#L862

KBaichoo commented on source/common/router/router.cc:
```diff
@@ -746,6 +746,14 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
+    wait_for_connect_ = true;
+    retry_connection_failure_buffer_limit_ =
+        std::max(retry_shadow_buffer_limit_, 2 * cluster_->perConnectionBufferLimitBytes());
+    callbacks_->setDecoderBufferLimit(retry_connection_failure_buffer_limit_);
```
we are now permanently increasing this buffer limit in this case, seems like we might want lower it back to its original value when we establish the connection. Alternatively, we could have a "connect buffer" where this is buffered in the case of connect and don't otherwise use this buffer

fl0Lec replied:
I would prefer not using a 3rd buffer, this would most likely have a none negligible impact on the memory and CPU performance due to the copy of the body until we get a connect.

I am fine with lowering the buffer limit to its original value once we get a connect.
I did not do it yet due to the recommendation in https://github.com/envoyproxy/envoy/blob/1b768b49f9a85a865c35735a4944a001975055e9/envoy/http/filter.h#L699-L701.

I guess this should be fine as the router filter should be the last stream filter in the filter chain. And that from a search in the code base, all filter call `setDecoderBufferLimit` in the `decodeHeaders` method.
I added a comment in https://github.com/envoyproxy/envoy/blob/main/envoy/http/filter.h.

KBaichoo replied:
There are custom filters that you might not see e.g. private that can call this during decodeBody, etc. 

Using a 3rd buffer isn't actually that bad because the underlying buffer object would just move around the slices (e.g. the backing store) so moving from the filter chain buffer to a connect buffer isn't that bad.

cc @alyssawilk  for additional thoughts

fl0Lec replied:
From my understanding the retry and shadow limit will most likely be equal to the FM buffer limit:
https://github.com/DataDog/envoy/blob/dfbcf58749b47bc8f98241d78023975da0b94f34/source/common/router/router.cc#L1000
```
retry_shadow_buffer_limit_ = callbacks_->decoderBufferLimit();
```

I do not think we can go above it see: https://github.com/envoyproxy/envoy/pull/32359#discussion_r1526455610
so we would need to increase it or did I missing something ?

KBaichoo reviewed (commented):
Thanks this looks much better.

/wait

fl0Lec reviewed (commented):
Thanks for the fast review, really appreciate it

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -746,6 +746,15 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
+    upstream_request->addUpstreamCallbacks(*this);
+    wait_for_connect_ = true;
+    retry_connection_failure_buffer_limit_ =
+        std::max(retry_shadow_buffer_limit_, 2 * cluster_->perConnectionBufferLimitBytes());
```
why 2x? Please comment inline
If you need to keep this as-is please handle and regression test overflow.

fl0Lec replied:
I already added a test: see ConnectionRetryBufferOverflow:
https://github.com/DataDog/envoy/blob/dfbcf58749b47bc8f98241d78023975da0b94f34/test/integration/integration_test.cc#L2473

alyssawilk replied:
why worry about reaching this limit before the connection is established?
If we're willing to buffer 5m upstream I don't think we should start out buffering 10M before connection establishment.

For overflow I meant integer overflow - I think if we keep the 2x you can actually overflow the uint32_t

fl0Lec replied:
That make sense, I did not thought of the integer overflow.  I will only increase it by 1MiB and  Checking that `cluster_->perConnectionBufferLimitBytes()<std::numeric_limits<uint32_t>::max()-1048576`

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1976,6 +2015,9 @@ void Filter::doRetry(bool can_send_early_data, bool can_use_http3, TimeoutRetry
   }
   UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && wait_for_connect_) {
```
this should probably be behind the runtime guard.  consider latching the runtime guard as a (temporary) member variable

fl0Lec replied:
done, it was not necessary here as wait_for_connect_ would have been false when creating the 1st requests.

alyssawilk commented on source/common/router/router.h:
```diff
@@ -471,6 +473,7 @@ class Filter : Logger::Loggable<Logger::Id::router>,
   bool downstreamEndStream() const override { return downstream_end_stream_; }
   uint32_t attemptCount() const override { return attempt_count_; }
   const std::list<UpstreamRequestPtr>& upstreamRequests() const { return upstream_requests_; }
+  void onUpstreamConnectionEstablished() override;
```
If this is from the new interface please comment so.

fl0Lec replied:
just to be sure do you want me to comment in the code that this is coming from the router interface or comment in this PR ?

alyssawilk commented on source/common/router/router.h:
```diff
@@ -589,15 +593,20 @@ class Filter : Logger::Loggable<Logger::Id::router>,
   bool downstream_response_started_ : 1;
   bool downstream_end_stream_ : 1;
   bool is_retry_ : 1;
+  bool wait_for_connect_ : 1;
```
I know we haven't been great about commenting this file but plase comment all new fields.

fl0Lec replied:
done

alyssawilk reviewed (commented):
Looks good!  Here's my first pass - I'm out tomorrow but I'll pick up again Monday
/wait

fl0Lec commented on test/integration/integration_test.cc:
```diff
@@ -2409,6 +2409,145 @@ TEST_P(IntegrationTest, RandomPreconnect) {
+  // Start Envoy and close the socket of the fake upstream server.
+  initialize();
+  Network::Address::InstanceConstSharedPtr upstream_address = fake_upstreams_[0]->localAddress();
+  fake_upstreams_[0]->closeSocket();
```
From what I tested this does not prevent the connection establishment, at least on my machine.
When running it with:
```
bzl test --test_arg="-l trace" //test/integration:integration_test --compilation_mode=fastbuild
```
I can see the connection being establish followed by a `read error: Resource temporarily unavailable, code: 0`

fl0Lec commented on source/common/router/router.cc:
```diff
@@ -837,6 +849,37 @@ void Filter::sendNoHealthyUpstreamResponse() {
+    // draining the buffer so we enable read on the downstream connection.
+    // This buffer is only used to send the body when retrying and shadowing which has been disable
+    // above.
+    callbacks_->modifyDecodingBuffer([](Buffer::Instance& data) { data.drain(data.length()); });
```
not sure we should actually drain the buffer as it might me used by some other filters.
But then should we keep the HCM buffer limit the new values

fl0Lec replied:
removed it, this was making quic and mobile test fail

alyssawilk commented:
hm, can you look into test failures? may need a main merge?
/wait

fl0Lec commented:
> hm, can you look into test failures? may need a main merge?
Rebase on merge, still need to figure out the new test failures
/wait

fl0Lec commented:
tests looks better now. Only one test failed not for arm64, will retest it
/retest

alyssawilk commented on source/common/router/router.h:
```diff
@@ -601,6 +613,15 @@ class Filter : Logger::Loggable<Logger::Id::router>,
+  // Set to true when the router is waiting to get an upcoming connection from the upstream,
+  // this will be set to true if ensure_connection_retry_ above is true and the router did
+  // not get a connection.
+  bool wait_for_connect_ : 1;
```
waiting_for_upstream_connection_

fl0Lec replied:
renamed it

alyssawilk commented on source/common/router/router.h:
```diff
@@ -297,14 +297,18 @@ using UpstreamRequestPtr = std::unique_ptr<UpstreamRequest>;
-                                               "envoy.reloadable_features.streaming_shadow")) {}
+        ensure_connection_retry_(
+            Runtime::runtimeFeatureEnabled("envoy.restart_features.ensure_connection_retry")),
+        wait_for_connect_(false), request_buffer_overflowed_(false),
```
why doesn't wait_for_connect_ start out as true?

fl0Lec replied:
I guess it make more sense to keep it disable until we create the 1st upstream make sure we have the conditions to wait for a connection.
But I don't have a strong opinion, I can also default it to true and disable it if we do not ensure connection retry

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -838,6 +848,27 @@ void Filter::sendNoHealthyUpstreamResponse() {
+void Filter::onUpstreamConnectionEstablished() {
+  ENVOY_LOG(trace, "Got connection with upstream.");
+  wait_for_connect_ = false;
+  if (!request_buffer_overflowed_ &&
```
why do we need a new boolean here?  Can't we just check if the retry state still exists?

fl0Lec replied:
I see the retry state get reset at multiple places. I guess it is more safe to use the already existing bool `request_buffer_overflowed_` which is only modify in `giveUpRetryAndShadow` to be sure we drop retry and shadow when we pass the retry_shadow_buffer_limit_

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -851,26 +882,36 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
+      // Increasing buffer limit if necessary.
+      if (callbacks_->decoderBufferLimit() <=
+          getLength(callbacks_->decodingBuffer()) + data.length()) {
+        callbacks_->setDecoderBufferLimit(retry_connection_failure_buffer_limit_);
```
this looks like it does a bunch of work each time we're over the limit.  Should we check the limit to only bump it up once?

fl0Lec replied:
I looks like from the latest test that we can not call setDecoderBufferLimit in that case where we are using `Http::AsyncStreamImpl`.

I therefor increase the buffer limit by default in the decodeHeader.
see failed test: https://github.com/envoyproxy/envoy/actions/runs/8538790525/job/23392163514

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -747,6 +747,16 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
                                         can_use_http3, false /*enable_half_close*/);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && ensure_connection_retry_) {
+    upstream_request->addUpstreamCallbacks(*this);
+    wait_for_connect_ = true;
```
don't we also need to set this at doRetry?
If we have a request with headers, get a connection, get a fast-fail, kick of a retry, this should be true again right?

fl0Lec replied:
I just wonder if we could have an issue in case of per try timeout:
1. request 1 per try timeout with a connection
2. router create request 2
3. request 1 call encodeHeaders

we would have `waiting_for_upstream_connection_ = true` when it would actually already have a connection with an upstream

My intention here was to ensure connection retry mostly until we get at least 1 connection. I believe we should us the `retry_shadow_buffer_limit_` for the other use case. 

Let me know if that make sense.

alyssawilk replied:
I still don't understand why we can't simply use the cluster limit - why do we need to add to it?

fl0Lec replied:
We can't use the cluster limit has the router might received more data than the cluster limit.
The upstream request will pause read when it has pass the buffer limit. So we need to keep a margin

alyssawilk commented on source/common/router/router.h:
```diff
@@ -601,6 +613,15 @@ class Filter : Logger::Loggable<Logger::Id::router>,
+  // mapped to the runtime guard "envoy.restart_features.ensure_connection_retry"
+  // on router creation.
+  bool ensure_connection_retry_ : 1;
+  // Set to true when the router is waiting to get an upcoming connection from the upstream,
```
is this actually set true when the router is waiting on a connection?  It looks like it only does under certain circumstances (retry configured)

fl0Lec replied:
That make sense, the comment was not clear at all, I updated it for 
```c++
  // Set to true when ensure_connection_retry_ above is true, retry on connection-failure is
  // enabled, and the router is waiting to get an upcoming connection from an upstream.
  // This will potentially increase the limit to which we are willing to buffer in order to ensure
  // we retry the connection failure.
```

alyssawilk reviewed (commented):
sorry for review lag - was out yesterday
here's another round of thoughts!
/wait

fl0Lec commented:
/retest

RyanTheOptimist commented:
@alyssawilk maintainer on-call ping

alyssawilk commented on source/common/router/router.h:
```diff
@@ -616,10 +616,11 @@ class Filter : Logger::Loggable<Logger::Id::router>,
+  // Set to true when ensure_connection_retry_ above is true, retry on connection-failure is
+  // enabled, and the router is waiting to get an upcoming connection from an upstream.
+  // This will potentially increase the limit to which we are willing to buffer in order to ensure
+  // we retry the connection failure.
```
so if the plan is to limit when this is set to not only be when it's waiting on upstream connection this isn't a good name.
What is this really supposed to be tracking? Is it the fact the buffer has an increased limit?  Let's name it for what it is.

fl0Lec replied:
Right, this is tracking the fact that we temporary increase the retry shadow buffer limit until we got a conneciton.
renamed it: `increased_retry_shadow_buffer_limit_until_got_connection_`

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -749,13 +749,21 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
+    // so it might goes above this limit.
+    uint32_t cluster_connection_buffer_limit = cluster_->perConnectionBufferLimitBytes();
+    // Ensure no integer overflow.
+    if (cluster_connection_buffer_limit < std::numeric_limits<uint32_t>::max() - 1048576) {
```
If we stick with +1m we could set to std::max (limit, limit +1m) rather than if/else and hoping the numbers are in sync.
I still don't understand why we can't use the per connection buffer limit though

fl0Lec replied:
done

alyssawilk reviewed (commented):
Pausing mid review to try to make sure I understand the goal of this PR
/wait-any

fl0Lec commented:
To clarify the goal of this PR:

The intent is to ensure we retry the connection-failure until we get at least one connection with an upstream.
While the router is waiting to get a connection with an upstream, we allow buffering up to cluster connection buffer limit plus 1 Mib.
Once a connection is acquired, we will use the default`retry_per_shadow_buffer_limit_` (min between the downstream connection buffer limit and route buffer limit).

**Why do we need plus 1 MiB**
We rely on the connection filter to pause reads on the downstream connection. Therefore, the router must be able to buffer more than the upstream buffer limit, as it will not pause read until the buffer is full (reached high watermark).

fl0Lec commented:
/retest

alyssawilk commented:
> To clarify the goal of this PR:
> 
> The intent is to ensure we retry the connection-failure until we get at least one connection with an upstream. While the router is waiting to get a connection with an upstream, we allow buffering up to cluster connection buffer limit plus 1 Mib. Once a connection is acquired, we will use the default`retry_per_shadow_buffer_limit_` (min between the downstream connection buffer limit and route buffer limit).
> 
> **Why do we need plus 1 MiB** We rely on the connection filter to pause reads on the downstream connection. Therefore, the router must be able to buffer more than the upstream buffer limit, as it will not pause read until the buffer is full (reached high watermark).

So in general, the router will buffer request data without sending a 4xx until it starts sending data upstream. After data has been sent upstream, the router will not be able to retry requests unless the request body fits in the shadow/retry buffer. 

All increasing the router buffer does is increase the amount of data the router will buffer in this manner before pushing back downstream.  This increases the risk of OOM so without a concrete value add I don't think it's worth tackling.  If you want larger buffers, you can configure larger buffers.  I don't think it's worth the additional complexity to already complicated code to reduce latency by buffering more if and only if we are waiting on upstream connections.  

Your e2e test passes just fine with a clean envoy checkout (after removing the reloadable flag set) so I think enovy handles so I think that shows envoy already does retries for large requests on connect failure?

bazel test --config=remote //test/integration:integration_test --test_timeout=120 --test_arg=--gtest_filter=IpVersionsAndHttp1Parser/IntegrationTest.EnsureConnectionRetry/IPv4BalsaParser

botengyao commented:
/assign @botengyao

fl0Lec commented:
> Fundamentally, the router used to be the point at which data was shipped upstream. So it would buffer the data in its own buffers until an upstream was selected, then send data upstream.
> I think a better fix would be to (optionally?) defer sending data upstream until the connection is established, trusting the HCM/watermark chain to keep the buffer small until we know we can safely proxy. If we had a connect failure we can in that one case retry using the HCM buffer instead of shadow/retry.

This all sounds more than reasonable to me. I would just have two concerns with this approach:
- waiting to get a callback from the upstream to inform the router that it has a connection might decrease Envoy performance. From what I understand now, once the upstream gets a connection, it can immediately start sending data. With this approach, it would have to fail for the data to be sent by the router and processed by the filters
- when relying on the HCM buffer plus watermark, the router will exceed the buffer watermark, which will pause reading on the downstream listener. Until then, everything is fine. 
But once it's started sending the data upstream, right now, the router copies the data, which makes sense when a none connection-failure retry is enabled (not when only connection-failure is enabled; maybe an optimization in the future). 
So the buffer will stay above the watermark. We can not really drain it because other filters might be using it. So what would be the path there?

I also have a question regarding the "HCM buffer instead of shadow/retry". From my understanding, they are the same: https://github.com/DataDog/envoy/blob/3763185bf4ce0fa6a3667fcd11921aa3edcdaf45/source/common/router/router.cc#L934.
Please bear with me has this is my first contribution in this repo.  

> but it adds a lot of complexity, and opens up an OOM vector for everyone who doesn't want a per-connection limit applied on a per-stream basis when upstream connections are slow to establish.

I agree that the workaround that I implemented introduces some memory risk and is not ideal. But I believe the discussion we are having now can bring great benefits.

botengyao commented on source/common/router/router.cc:
```diff
@@ -850,16 +883,17 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
   if (buffering &&
-      getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_) {
+      ((!increased_retry_shadow_buffer_limit_until_got_connection_ &&
+        getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_) ||
```
Let's take a step back for your first [commit #1](https://github.com/envoyproxy/envoy/pull/32359/commits/29a8867730776f031caec097e295e74e72d78998) with explicitly calling `onDecoderFilterAboveWriteBufferHighWatermark`.

I am actually looking into how we are buffering right now from this code path, taking a look at this comment: https://github.com/envoyproxy/envoy/blob/b47fc6648d7c2dfe0093a601d44cb704b7bad382/source/common/router/router.cc#L901-L907, it states we don't need to worry about the 413 since we early give up if the data is over-limit here. 

I doubt we probably can continue buffer the data if `getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_)` once, and it is actually using streaming mode with  `state_.decoder_filters_streaming_ = true`, and the watermark can be applied without worrying 413 in filter_manager and `onDecoderFilterAboveWriteBufferHighWatermark` is called already when we are moving data to the `requests_buffered_data` IIUC
https://github.com/envoyproxy/envoy/blob/b47fc6648d7c2dfe0093a601d44cb704b7bad382/source/common/http/filter_manager.cc#L542-L547, this is just like how watermark buffer works: "add data at first and then check the watermark".  cc @mattklein123 who added this. 

In this case we don't need to adjust the downstream decoder buffer limit, but simply to let this code go through when we are waiting for connection.

@fl0Lec WDYT and do you wanna give this a shot?

botengyao replied:
> Sorry was OOO,
> 
> I would also rely on the buffer reaching the watermark to pause the read. So:
> 
> * getting connection from downstream
> * trying to get a connection with upstream
> * filter buffer gets full (we go over the watermark which pauses the downstream read)
> * we get a connection
> 
> At that point to go bellow the watermark, we have the following options:
> 
> 1. draining the buffer: no possible as other filter might also use it.
> 2. increasing the limit to be right above the currently buffered data. Probably the option for now
> 3. do we have other option ?
> 
> WDYT @botengyao ?

What is the specific reason that we cannot drain the buffer since router is the terminated filter?

Note, we may cannot drain here if we have shadow policy since the data will also be used for shadow, the draining should be conditional. Eventually this can be drained if router will not use the data anymore/already shipped to upstream, but this looks like not today's case. cc @alyssawilk.

botengyao replied:
thanks, then we need to re-adjust the buffer limit with the current buffer length with a hard limit to drop with this approach. This delays the adjustment of buffer limit but still needs to adjust.

fl0Lec replied:
That makes sense, I implemented it in the latest commit (0828ffd19616a2cf02903ba4426fbd2aa93d0453). Let's see if all test pass

botengyao reviewed (commented):
Thanks @fl0Lec for the contribution! sry for the delay and it actually takes me more time than I expected to understand this PR history ;-)

Left one comment to kick off the discussion.

botengyao commented:
>Aside, @botengyao should the fix here be returning StopIterationAndWatermark rather than sendLocalReply? Seems if we've never sent data upstream, we should just delay until we have one.
https://github.com/envoyproxy/envoy/commit/3c234770d2b1660fa6f3fdb1f9c3916933ec7b84#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910

I don't think so, since retry & shadow is disabled, and upstream_request is created during decoding header, the empty here means no more upstream_requests should happen, so we can local reply.

adisuissa commented:
/wait

fl0Lec commented:
> > Aside, @botengyao should the fix here be returning StopIterationAndWatermark rather than sendLocalReply? Seems if we've never sent data upstream, we should just delay until we have one.
> > [3c23477#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910](https://github.com/envoyproxy/envoy/commit/3c234770d2b1660fa6f3fdb1f9c3916933ec7b84#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910)
> 
> I don't think so, since retry & shadow is disabled, and upstream_request is created during decoding header, the empty here means no more upstream_requests should happen, so we can local reply.

From my understanding empty means that we are waiting for a retry to happens (we are in the backoff wait).
I believe this issue is linked to it https://github.com/envoyproxy/envoy/issues/15211.
Ideally we would wait for the retry to happened instead of returning a 507

fl0Lec commented:
Sorry was OOO, 

I would also rely on the buffer reaching the watermark to pause the read. So:
- getting connection from downstream
- trying to get a connection with upstream
- filter buffer get full (we go over the watermark which pause the downstream read)
- we get a connection

At that point to go bellow the watermark, we have the following options:
1. draining the buffer: no possible as other filter might also use it.
2. increasing the limit to be right above the currently buffered data. Probably the option for now
3. do we have other option ?

WDYT @botengyao

fl0Lec commented:
/remove-wait

botengyao commented:
>
Aside, @botengyao should the fix here be returning StopIterationAndWatermark rather than sendLocalReply? Seems if we've never sent data upstream, we should just delay until we have one.
[3c23477#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910](https://github.com/envoyproxy/envoy/commit/3c234770d2b1660fa6f3fdb1f9c3916933ec7b84#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910)

I don't think so, since retry & shadow is disabled, and upstream_request is created during decoding header, the empty here means no more upstream_requests should happen, so we can local reply.

From my understanding empty means that we are waiting for a retry to happens (we are in the backoff wait).
I believe this issue is linked to it https://github.com/envoyproxy/envoy/issues/15211.
Ideally we would wait for the retry to happened instead of returning a 507



> > > Aside, @botengyao should the fix here be returning StopIterationAndWatermark rather than sendLocalReply? Seems if we've never sent data upstream, we should just delay until we have one.
> > > [3c23477#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910](https://github.com/envoyproxy/envoy/commit/3c234770d2b1660fa6f3fdb1f9c3916933ec7b84#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR910)
> > 
> > 
> > I don't think so, since retry & shadow is disabled, and upstream_request is created during decoding header, the empty here means no more upstream_requests should happen, so we can local reply.
> 
> From my understanding empty means that we are waiting for a retry to happens (we are in the backoff wait). I believe this issue is linked to it #15211. Ideally we would wait for the retry to happened instead of returning a 507

This is not the same issue I believe and let's go back to this PR ;-)

botengyao reviewed (commented):
Thanks for working on this! Looks like a challenging problem.

botengyao commented:
So, the goal of this PR is going to allow retry if the downstream decoder buffer is full when waiting for the upstream connection (with a hard limit to avoid super large requests as well). This is basically is to add the flow control for the retry buffer, but the key point here is we don't drain the downstream decoder buffer that is used for retry even some retry conditions are no longer held.

This PR discussed 5 approaches from my understanding.
1. Explicitly call the `onAboveWriteBufferHighWaterMark` and read enable it when the upstream connection established, as [commit #1](https://github.com/envoyproxy/envoy/pull/32359/commits/29a8867730776f031caec097e295e74e72d78998). However, we should never let the data be more than the downstream decoder buffer limit, since we will never drain it, and the data will continue to disable the read if we call `addDecodedData(data, true)`.

2. Using the upstream flow control by adjusting/incresing the downstream filter buffer limit to `upstream_buffer_limit + 1Mb` or whatever.

3. Discussed above like, leverage the downstream decoder buffer flow control introduced by `addDecodeData(data, true)` to let the retry buffer exceed the buffer limit once. Since we cannot drain the buffer, to re-read enable the downstream read we need to change the buffer limit to `current_data.length()`. If it is too large we also need to drop. This delayed the buffer adjustment from approach 2.

4. As Allyssa mentioned, in high level another approach is do not buffer data before upstream connection is established with the HCM buffer and watermark chain. Alyssa, are you suggesting the network buffer/codec buffer from HCM, or the decoder filter chain buffer? If it is the later then it is similar as approach 3 in high level.

5. Maybe a third buffer to only buffer the connection failure retry, and drain it when upstream connection is established. The cost should be small with only buffer slice movement, cc @KBaichoo.

Approach 3 is what I think to solve the problem, but still with the buffer limit change. Otherwise 4/5 deserves exploration and further tests as @fl0Lec brought up. Happy to hear other thoughts @alyssawilk @KBaichoo.

fl0Lec reviewed (commented):
Thanks for you review and feedback. 
The latest approach is to allows going over the buffer limit (which will pause the read on downstream) up to the limit defined at the listener level + 1 Mib.
Once Envoy got a connection, it drop retry and increase the buffer limit to its current length which un-pause the read on the downstream listener.

fl0Lec commented:
/retest

phlax commented:
@fl0Lec it seems ci is not running due to conflicts (altho not clear what conflict)

could you merge main please

/wait

fl0Lec commented:
/retest

fl0Lec commented:
> @fl0Lec it seems ci is not running due to conflicts (altho not clear what conflict)
> 
> could you merge main please
> 

Done, all tests passed 👍

ravenblackx commented:
Ping @botengyao

botengyao commented on test/integration/integration_test.cc:
```diff
@@ -2427,6 +2427,97 @@ TEST_P(IntegrationTest, RandomPreconnect) {
+  // the new upstream before the retry.
+  config_helper_.addRuntimeOverride("upstream.base_retry_backoff_ms", "200");
+  // enable feature flags envoy.restart_features.ensure_connection_retry
+  config_helper_.addRuntimeOverride("envoy.restart_features.ensure_connection_retry", "true");
```
Could you also add a test to verify other retry features together with connection retries?

fl0Lec replied:
Done, I added some test based on 5xx retry with the feature enable:
1. test that we still retry 5xx when body is smaller than limit
2. test that we don't retry 5xx when body is bigger than limit

botengyao commented on source/common/router/router.cc:
```diff
@@ -846,6 +857,30 @@ void Filter::sendNoHealthyUpstreamResponse() {
+    // Disabling retry and shadow: the request is to big.
+    giveUpRetryAndShadow();
+    if (getLength(callbacks_->decodingBuffer()) > callbacks_->decoderBufferLimit()) {
+      callbacks_->setDecoderBufferLimit(getLength(callbacks_->decodingBuffer()));
```
So the new bufferlimit right now is  < `retry_connection_failure_buffer_limit_` and >`retry_shadow_buffer_limit_`, other than that we will early local reply later, am I reading correctlly?

fl0Lec replied:
that is correct

botengyao commented on test/integration/integration_test.cc:
```diff
@@ -2427,6 +2427,97 @@ TEST_P(IntegrationTest, RandomPreconnect) {
+  auto& encoder = encoder_decoder.first;
+  auto& response = encoder_decoder.second;
+  // Send more data than the buffer limit.
+  codec_client_->sendData(encoder, 256, true);
```
Do you also want to simulate multiple read situations?

botengyao replied:
For multiple reads, it means the data from the buffer can be through multiple reads from I/O and then the buffer limit can be triggered. Did you try this scenario?

botengyao reviewed (commented):
Thanks!

botengyao commented:
/wait

fl0Lec reviewed (commented):
Sorry got caught in other project. I merge main and added some test. I did not managed to make multiple read work yet.

fl0Lec commented:
/retest

htuch commented:
@botengyao friendly ping.

nezdolik commented:
friendly ping @botengyao :)

botengyao reviewed (commented):
Thanks for working on this!  Overall LGTM with this approach module a test scenario. 

Let's take a step back, the value right now is we slightly increase the tolerate buffer limit with max of shadow buffer limit and cluster's per connection buffer limit + 1MB for the connection give-up. for your original problem, could it be resolved by adjusting the configurations?

We also have other approaches described in
https://github.com/envoyproxy/envoy/pull/32359#issuecomment-2083643853, but need more explorations, and willing to hear @KBaichoo, @alyssawilk's input.

botengyao commented:
/wait
